From 8c84c821536632c04061545e285afadc59f80120 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sun, 28 Mar 2021 14:58:01 -0500
Subject: [PATCH 057/196] Add ScriptExecutionData and thread it through
 EvalScript. (#101)

This is required to collect the data for the BIP341 sighash, which, among other things, contains the last executed codeseparator position.
---
 src/bench/verify_script.cpp       |  3 +-
 src/script/interpreter.cpp        | 18 +++++--
 src/script/interpreter.h          |  8 +++-
 src/script/script_exec_data.h     | 23 +++++++++
 src/test/sigcheckcount_tests.cpp  |  9 ++--
 src/test/sighash_bip341_tests.cpp | 78 +++++++++++++++++++++++++++++++
 6 files changed, 128 insertions(+), 11 deletions(-)
 create mode 100644 src/script/script_exec_data.h

diff --git a/src/bench/verify_script.cpp b/src/bench/verify_script.cpp
index 30c91154d..6dc72d2e6 100644
--- a/src/bench/verify_script.cpp
+++ b/src/bench/verify_script.cpp
@@ -31,9 +31,10 @@ static void VerifyNestedIfScript(benchmark::Bench &bench) {
     bench.run([&] {
         auto stack_copy = stack;
         ScriptExecutionMetrics metrics = {};
+        ScriptExecutionData execdata = {};
         ScriptError error;
         bool ret = EvalScript(stack_copy, script, 0, BaseSignatureChecker(),
-                              metrics, &error);
+                              metrics, execdata, &error);
         assert(ret);
     });
 }
diff --git a/src/script/interpreter.cpp b/src/script/interpreter.cpp
index 9ce2227b3..0d122c387 100644
--- a/src/script/interpreter.cpp
+++ b/src/script/interpreter.cpp
@@ -214,7 +214,8 @@ static bool EvalChecksig(const valtype &vchSig, const valtype &vchPubKey,
 
 bool EvalScript(std::vector<valtype> &stack, const CScript &script,
                 uint32_t flags, const BaseSignatureChecker &checker,
-                ScriptExecutionMetrics &metrics, ScriptError *serror) {
+                ScriptExecutionMetrics &metrics, ScriptExecutionData &execdata,
+                ScriptError *serror) {
     static const CScriptNum bnZero(0);
     static const CScriptNum bnOne(1);
     static const valtype vchFalse(0);
@@ -232,10 +233,12 @@ bool EvalScript(std::vector<valtype> &stack, const CScript &script,
         return set_error(serror, ScriptError::SCRIPT_SIZE);
     }
     int nOpCount = 0;
+    uint32_t opcode_pos = 0;
+    execdata.m_codeseparator_pos = 0xffff'ffff;
     constexpr bool fRequireMinimal = true;
 
     try {
-        while (pc < pend) {
+        for (; pc < pend; ++opcode_pos) {
             bool fExec = vfExec.all_true();
 
             //
@@ -926,6 +929,7 @@ bool EvalScript(std::vector<valtype> &stack, const CScript &script,
                     case OP_CODESEPARATOR: {
                         // Hash starts after the code separator
                         pbegincodehash = pc;
+                        execdata.m_codeseparator_pos = opcode_pos;
                     } break;
 
                     case OP_CHECKSIG:
@@ -1939,16 +1943,19 @@ bool VerifyScript(const CScript &scriptSig, const CScript &scriptPubKey,
     }
 
     ScriptExecutionMetrics metrics = {};
+    ScriptExecutionData execdata = {};
 
     // scriptSig and scriptPubKey must be evaluated sequentially on the same
     // stack rather than being simply concatenated (see CVE-2010-5141)
     std::vector<valtype> stack, stackCopy;
-    if (!EvalScript(stack, scriptSig, flags, checker, metrics, serror)) {
+    if (!EvalScript(stack, scriptSig, flags, checker, metrics, execdata,
+                    serror)) {
         // serror is set
         return false;
     }
     stackCopy = stack;
-    if (!EvalScript(stack, scriptPubKey, flags, checker, metrics, serror)) {
+    if (!EvalScript(stack, scriptPubKey, flags, checker, metrics, execdata,
+                    serror)) {
         // serror is set
         return false;
     }
@@ -1978,7 +1985,8 @@ bool VerifyScript(const CScript &scriptSig, const CScript &scriptPubKey,
         CScript pubKey2(pubKeySerialized.begin(), pubKeySerialized.end());
         popstack(stack);
 
-        if (!EvalScript(stack, pubKey2, flags, checker, metrics, serror)) {
+        if (!EvalScript(stack, pubKey2, flags, checker, metrics, execdata,
+                        serror)) {
             // serror is set
             return false;
         }
diff --git a/src/script/interpreter.h b/src/script/interpreter.h
index 4a0a297d6..856bafbd5 100644
--- a/src/script/interpreter.h
+++ b/src/script/interpreter.h
@@ -9,6 +9,7 @@
 
 #include <primitives/transaction.h>
 #include <script/script_error.h>
+#include <script/script_exec_data.h>
 #include <script/script_flags.h>
 #include <script/script_metrics.h>
 #include <script/sighashtype.h>
@@ -84,13 +85,16 @@ using MutableTransactionSignatureChecker =
 
 bool EvalScript(std::vector<std::vector<uint8_t>> &stack, const CScript &script,
                 uint32_t flags, const BaseSignatureChecker &checker,
-                ScriptExecutionMetrics &metrics, ScriptError *error = nullptr);
+                ScriptExecutionMetrics &metrics, ScriptExecutionData &execdata,
+                ScriptError *error = nullptr);
 static inline bool EvalScript(std::vector<std::vector<uint8_t>> &stack,
                               const CScript &script, uint32_t flags,
                               const BaseSignatureChecker &checker,
                               ScriptError *error = nullptr) {
     ScriptExecutionMetrics dummymetrics;
-    return EvalScript(stack, script, flags, checker, dummymetrics, error);
+    ScriptExecutionData dummyexecdata;
+    return EvalScript(stack, script, flags, checker, dummymetrics,
+                      dummyexecdata, error);
 }
 
 /**
diff --git a/src/script/script_exec_data.h b/src/script/script_exec_data.h
new file mode 100644
index 000000000..c7fe847e6
--- /dev/null
+++ b/src/script/script_exec_data.h
@@ -0,0 +1,23 @@
+// Copyright (c) 2021 The Logos Foundation
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_SCRIPT_SCRIPT_EXEC_DATA_H
+#define BITCOIN_SCRIPT_SCRIPT_EXEC_DATA_H
+
+#include <uint256.h>
+
+/**
+ * Struct for holding data generated during script execution which is used
+ * during script execution, e.g. for computing sig hashes.
+ */
+struct ScriptExecutionData {
+    /**
+     * Opcode position of the last executed OP_CODESEPARATOR.
+     * 
+     * This allows signatures to commit to certain code paths.
+     */
+    uint32_t m_codeseparator_pos;
+};
+
+#endif // BITCOIN_SCRIPT_SCRIPT_EXEC_DATA_H
diff --git a/src/test/sigcheckcount_tests.cpp b/src/test/sigcheckcount_tests.cpp
index c1b0eefc0..b92559e99 100644
--- a/src/test/sigcheckcount_tests.cpp
+++ b/src/test/sigcheckcount_tests.cpp
@@ -114,9 +114,10 @@ static void CheckEvalScript(const stacktype &original_stack,
         ScriptError err = ScriptError::UNKNOWN;
         stacktype stack{original_stack};
         ScriptExecutionMetrics metrics;
+        ScriptExecutionData execdata;
 
-        bool r =
-            EvalScript(stack, script, flags, dummysigchecker, metrics, &err);
+        bool r = EvalScript(stack, script, flags, dummysigchecker, metrics,
+                            execdata, &err);
         BOOST_CHECK(r);
         BOOST_CHECK_EQUAL(err, ScriptError::OK);
         BOOST_CHECK(stack == expected_stack);
@@ -211,9 +212,11 @@ BOOST_AUTO_TEST_CASE(test_evalscript) {
     {
         stacktype stack{txsigschnorr};
         ScriptExecutionMetrics metrics;
+        ScriptExecutionData execdata;
         metrics.nSigChecks = 12345;
         bool r = EvalScript(stack, CScript() << pub << OP_CHECKSIG,
-                            SCRIPT_ENABLE_SIGHASH_FORKID, dummysigchecker, metrics);
+                            SCRIPT_ENABLE_SIGHASH_FORKID, dummysigchecker,
+                            metrics, execdata);
         BOOST_CHECK(r);
         BOOST_CHECK_EQUAL(metrics.nSigChecks, 12346);
     }
diff --git a/src/test/sighash_bip341_tests.cpp b/src/test/sighash_bip341_tests.cpp
index 420ab175d..8241c1815 100644
--- a/src/test/sighash_bip341_tests.cpp
+++ b/src/test/sighash_bip341_tests.cpp
@@ -4,7 +4,10 @@
 
 #include <chain.h>
 #include <coins.h>
+#include <core_io.h>
 #include <hash.h>
+#include <policy/policy.h>
+#include <script/interpreter.h>
 #include <script/script.h>
 #include <script/standard.h>
 
@@ -12,6 +15,9 @@
 
 #include <boost/test/unit_test.hpp>
 
+typedef std::vector<uint8_t> valtype;
+typedef std::vector<valtype> stacktype;
+
 BOOST_FIXTURE_TEST_SUITE(sighash_bip341_tests, BasicTestingSetup)
 
 BOOST_AUTO_TEST_CASE(prepare_spent_outputs) {
@@ -97,4 +103,76 @@ BOOST_AUTO_TEST_CASE(precompute_bip341_hashes) {
                                "a5345b6977e89ea3c"));
 }
 
+static const std::vector<uint32_t> allflags{
+    SCRIPT_VERIFY_NONE,
+    STANDARD_SCRIPT_VERIFY_FLAGS,
+};
+
+void CheckCodesepPos(const CScript &script,
+                     const uint32_t expected_codesep_pos) {
+    for (uint32_t flags : allflags) {
+        BaseSignatureChecker sigchecker;
+        ScriptExecutionMetrics metrics = {};
+        ScriptExecutionData execdata = {};
+        stacktype stack = {};
+        bool r =
+            EvalScript(stack, script, flags, sigchecker, metrics, execdata);
+        BOOST_CHECK(r);
+        BOOST_CHECK_EQUAL(execdata.m_codeseparator_pos, expected_codesep_pos);
+        BOOST_CHECK_MESSAGE(execdata.m_codeseparator_pos ==
+                                expected_codesep_pos,
+                            "For script '" << ScriptToAsmStr(script) << "'");
+    }
+}
+
+BOOST_AUTO_TEST_CASE(script_execution_data) {
+    valtype data10(10);
+    valtype data520(520);
+    // Test unconditional cases
+    CheckCodesepPos(CScript() << OP_1, 0xffff'ffff);
+    CheckCodesepPos(CScript() << data10 << OP_1, 0xffff'ffff);
+    CheckCodesepPos(CScript() << data520 << OP_1, 0xffff'ffff);
+    CheckCodesepPos(CScript() << OP_CODESEPARATOR << data10 << OP_1, 0);
+    CheckCodesepPos(CScript() << OP_CODESEPARATOR << data520 << OP_1, 0);
+    CheckCodesepPos(CScript() << data520 << OP_CODESEPARATOR << OP_1, 1);
+    CheckCodesepPos(CScript() << data10 << OP_CODESEPARATOR << OP_1, 1);
+    CheckCodesepPos(CScript() << data520 << OP_1 << OP_CODESEPARATOR, 2);
+    CheckCodesepPos(CScript() << data520 << data10 << OP_1 << OP_CODESEPARATOR,
+                    3);
+    CheckCodesepPos(CScript() << data520 << OP_NOP << OP_NOP << OP_NOP << OP_NOP
+                              << OP_NOP << OP_CODESEPARATOR,
+                    6);
+
+    // Test conditional cases
+    CheckCodesepPos(CScript() << 0 << OP_IF << OP_NOP << OP_CODESEPARATOR
+                              << OP_NOP << OP_ENDIF,
+                    0xffff'ffff);
+    CheckCodesepPos(CScript() << 1 << OP_IF << OP_NOP << OP_CODESEPARATOR
+                              << OP_NOP << OP_ENDIF,
+                    3);
+    CheckCodesepPos(CScript() << 0 << OP_NOTIF << OP_NOP << OP_CODESEPARATOR
+                              << OP_NOP << OP_ENDIF,
+                    3);
+    CheckCodesepPos(CScript() << 1 << OP_NOTIF << OP_NOP << OP_CODESEPARATOR
+                              << OP_NOP << OP_ENDIF,
+                    0xffff'ffff);
+    CheckCodesepPos(CScript()
+                        << 1 << 1 << 1 << OP_IF << OP_IF << OP_IF
+                        << OP_CODESEPARATOR << OP_ENDIF << OP_ENDIF << OP_ENDIF,
+                    6);
+    CheckCodesepPos(CScript()
+                        << 1 << 0 << 1 << OP_IF << OP_IF << OP_IF
+                        << OP_CODESEPARATOR << OP_ENDIF << OP_ENDIF << OP_ENDIF,
+                    0xffff'ffff);
+    CheckCodesepPos(CScript() << 1 << 0 << 1 << OP_IF << OP_IF << OP_IF
+                              << OP_CODESEPARATOR << OP_ENDIF << OP_ELSE
+                              << OP_CODESEPARATOR << OP_ENDIF << OP_ENDIF,
+                    9);
+    CheckCodesepPos(CScript() << 1 << 0 << 1 << OP_IF << OP_CODESEPARATOR
+                              << OP_IF << OP_IF << OP_CODESEPARATOR << OP_ENDIF
+                              << OP_ELSE << OP_CODESEPARATOR << OP_ENDIF
+                              << OP_ELSE << OP_CODESEPARATOR << OP_ENDIF,
+                    10);
+}
+
 BOOST_AUTO_TEST_SUITE_END()
-- 
2.39.5 (Apple Git-154)

